Spring Week7 인가(Authorization) 구현하기

Flying-Fisherman opened the pull request:
Spring Security 7-1강 내용을 따라가고 있습니다.

Flying-Fisherman committed:
Spring Week7 인가(Authorization) 구현하기

Spring Security 7-1강

johngrib reviewed (commented):
이번주 부터는 커밋 메시지도 같이 연습해 봅시다. 커밋 메시지 타이틀과 커밋 메시지 본문을 한국어로 작성해서 올려주시면 커밋 메시지에 대해서도 리뷰해 드릴게요.

johngrib reviewed (commented):
099038855cdedbacbacebd495482396cbf01acff

![image](https://user-images.githubusercontent.com/1855714/121892319-8dd00080-cd57-11eb-8284-1ce444f6ce1e.png)

이 커밋은 수업 과정을 따라하면서 어느 과정을 학습했는지 남기는 기록에 가깝습니다. 이제부터는 어떤 작업을 했는지를 확실히 구분해서 작업에 따라 커밋을 작성해보시는 것을 권해드리고 싶어요.

johngrib commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,33 @@
+
+import javax.servlet.Filter;
+
+@Configuration
```
`@Configuration` 애노테이션의 JavaDoc을 읽고 학습해 보세요, 그리고 어떤 내용이었는지 글로 작성해 보세요. 애노테이션을 사용할 때마다 내가 저 애노테이션의 JavaDoc을 읽은 적이 있는지 자기 자신에게 꼭 되물어 보세요.

Flying-Fisherman replied:
`@Configuration` 어노테이션의 JavaDoc을 읽어봤습니다.

생각보다 많은 활용처가 있어서 아직은 일부분의 사용법 만 이해했습니다.

`@Configuration` 은 클래스에서의 환경구성을 돕는 어노테이션으로, `Configuration` 인터페이스를 구현하여 정의합니다.

`@Configuration` 어노테이션을 구현함으로써 하나 이상의 `@Bean` 메서드를 선언한 클래스를 나타내고, 이러한 정의된 Bean들의 정의와 서비스 요청을 `Spring Container`에 생성하는 처리를 해줍니다.

이 전에 Spring에서는 싱글턴 방식(인스턴스가 오직 1개만 생성되어야 하는 경우의 패턴)으로 스프링 컨테이너에  `Bean`이 등록되며, 이 `Bean`들을 등록하는데는 xml에 서술하여 일일이 등록해야 합니다.
즉, 사용하는 `Bean`이 많아질 수록 등록 및 관리에 많은 시간을 할애하게 됩니다.

하지만 `@Configuration`어노테이션을 명시함으로써 해당 클래스에서 1개 이상의 Bean을 생성하고 있음을 명시하고, 생산력 저하를 방지한다.

`@Configuration`없이 `@Bean`만 사용해도 스프링 빈으로 등록이 되나, 메소드 호출을 통해 객체를 생성할 싱글톤을 보장하지 못한다. 그렇기 때문에 Spring 설정 정보를 위해서는 반드시 `@Configuration`을 사용해주어야 한다.

비슷한 어노테이션으로 `@Component`어노테이션이 있는데, 이 둘의 차이점으로는 `@Configuration`은 스프링 컨테이너(스프링 컨텍스트)에 등록되어 있는 Bean을 반환하지만, `@Component` 어노테이션은 새로 생성된 Bean이 반환된다.

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFliter.java:
```diff
@@ -0,0 +1,64 @@
+        }
+
+        String method = request.getMethod();
+        if (method.equals("GET")) {
+            return true;
+        }
+
+        if (method.equals("OPTIONS")) {
```
`GET`, `OPTIONS` 같은 것들은 enum으로 어디엔가 정의된 게 있지 않을까요?

Flying-Fisherman replied:
이제 이 부분에 대해서 제가 저번주 과제풀이 영상을 못봐서 다시 확인하고 있습니다 ㅠㅠ

johngrib replied:
과제풀이를 보고 찾지 마시고 다른 곳에서 한번 찾아 보세요. Spring에 있는 `HttpMethod` 같은 것들을 사용할 수 있겠죠.

Flying-Fisherman committed:
유저 로그인 시 권한 부여 및 통제 설정 / 유저 ID에 따른 권한 설정

Controller에 존재하는 각 메서드에 대한 유저들의 권한 여부를 확인하는 기초 작업을 실시
UserId에 따른 권한 식별에 대한 기초 작업 실시

johngrib reviewed (commented):
![image](https://user-images.githubusercontent.com/1855714/122218172-1f22ac80-cee9-11eb-87b5-11da7d37bec3.png)
4d2adcc943d5d630e1f206f508363f03a0aeef9c 커밋 메시지도 그렇고 변경사항도 읽어보면 커밋을 두 개로 분리하는 것이 더 좋았을 것 같아요.

johngrib commented on app/build.gradle:
```diff
@@ -65,6 +65,9 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-validation'
 
+    // Spring Security
+    implementation 'org.springframework.boot:spring-boot-starter-security'
```
스프링 시큐리티에 대해 소개해 주세요.
시간이 충분하시다면 다음 문서가 도움이 될 거에요.

johngrib commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,42 @@
+        Filter authenticationErrorFilter = new AuthenticationErrorFilter();
+
+        http
+                .csrf().disable()
+                .addFilter(authenticationFilter)
+                .addFilterBefore(authenticationErrorFilter,
+                        JwtAuthenticationFliter.class)
+                .sessionManagement()
+                    .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+                .and()
+                .exceptionHandling()
+                    .authenticationEntryPoint(new HttpStatusEntryPoint(HttpStatus.UNAUTHORIZED));
```
이 각 라인들이 어떤 의미를 갖고 있는지 한 줄 한 줄 설명해 보세요.

Flying-Fisherman replied:
`.csrf().disable()` => `@EnableWebSecurity`가 활성화 되어 있는 경우(위의 `@EnableGlobalMethSecurity`도 포함) 자동적으로 CSRF 프로텍션 기능(페이지)을 활성화합니다.
일반적인 사용인 경우에는 명시적으로 보여주기 위해 작성하지만, 사용하지 않기 위해 `.disable()`을 통해 비활성화합니다.

`.addFilter()`, `.addFilterBefore()` => `.addFilter()`을 통해 커스텀 필터를 지정합니다.
`.addFilterBefore(A, B.class)`를 통해 B.class 이전에 A 필터를 올 수 있도록 지정할 수 있습니다.

`.sessionManagement()` => 세션에 대한 강제성(?)을 부여합니다. 즉, 동일한 계정으로 접속이 허용되는 세션을 어떻게 통제할 지를 표현합니다.
`.sessionCreationPolicy()` => 세션에 대한 정책을 명시합니다. 우리는 JWT 토큰 방식을 사용하기 때문에 Stateless로 정책을 명시합니다.

`.and()` => http에 대한 여러가지 작업을 연결해주는 역할을 합니다.

`.exceptionHandling()` => 발생하는 예외처리를 다룰 수 있는 Configurer 역할을 수행합니다.

`.authenticationEntryPoint()` => EntryPoint를 어떻게 지정할지에 대한 설정을 합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFliter.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain) throws IOException, ServletException {
```
메소드 시그니처가 너무 길어서 2줄이나 3줄로 개행해줘도 좋을 것 같아요.

Flying-Fisherman committed:
Password 암호화 / 복호화 구현

BCryptPasswordEncoder을 이용한 암호화/복호화 구현
암호화/복호화 사용으로 인한 각 Test 및 Service 수정

johngrib reviewed (commented):
![image](https://user-images.githubusercontent.com/1855714/122393236-a20e3a80-cfaf-11eb-95fb-fc6d38d8c2f9.png)

커밋 메시지를 작성할 때 다음과 같은 시도를 해 보세요.

- 문장을 완성합니다.
- 후배에게 설명한다고 생각하고 친절하게 작성해 보세요.

Flying-Fisherman committed:
Controller 및 Service 일부 수정

`AuthenticationService`에서 인코딩 부분에서의 하드코딩 부분을 수정했습니다.

`JwtAuthenticationFilter`에서의 메소드 시그니처를 개행했습니다.

`ProductController`에서의 `@RequestAttribute`를 비활성화(태그)처리 했습니다.
-> 차후 권한 부여 관련으로 사용 연구 예정입니다.

`SecurityJavaConfig`에서의 `/h2-console`에 대한 Spring-Security 에서의 예외처리를 했습니다.

`UserController`에서의 인증여부 확인을 추가하였습니다.

이외의 추가적인 테스트 수정 및 추가하였습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/ProductController.java:
```diff
@@ -40,16 +41,19 @@ public Product detail(@PathVariable Long id) {
 
     @PostMapping
     @ResponseStatus(HttpStatus.CREATED)
+//    @PreAuthorize("isAuthenticated() and hasAuthority('USER')")
```
이런 주석은 그냥 삭제하는 것이 좋겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/AuthenticationErrorFilter.java:
```diff
@@ -0,0 +1,26 @@
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class AuthenticationErrorFilter extends HttpFilter {
```
필터의 역할에 대해 JavaDoc으로 주석을 남겨 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFliter.java:
```diff
@@ -0,0 +1,68 @@
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class JwtAuthenticationFliter extends BasicAuthenticationFilter {
```
이 필터의 역할과 책임을 JavaDoc으로 작성해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/filters/JwtAuthenticationFliter.java:
```diff
@@ -0,0 +1,68 @@
+
+    private boolean filterWithPathAndMethod(HttpServletRequest request) {
+        String path = request.getRequestURI();
+        if (!path.startsWith("/products")) {
```
이 조건이 들어간 이유는 무엇인가요?

Flying-Fisherman committed:
인증 부분 수정, 인증 관련 각 메서드 수정

`JwtAuthenticationFliter`에서의 중간처리 중 `Users`도 추가하였습니다.

`JwtUtil`에서의 하드코딩이 된 부분을 수정하였습니다.

`UserAuthentication`에서 ADMIN 권한을 추가할 수 있도록 했습니다.

`UserController`에서의 권한 확인 및 전달자 부분을 일부 수정했습니다.

`UserService`에서 업데이트 시의 권한 확인 및 예외처리를 추가하였습니다.\

이 외 확인 테스트 일부 수정하였습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -16,4 +18,9 @@ public static void main(String[] args) {
+
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder();
```
bcrypt에 대해서도 조사해서 설명해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -29,10 +34,17 @@ public User registerUser(UserRegistrationData registrationData) {
-    public User updateUser(Long id, UserModificationData modificationData) {
+    public User updateUser(Long id, UserModificationData modificationData, Long userId) {
+        if(id != userId) {
+            throw new AccessDeniedException("Access Denied");
```
이 부분의 에러 메시지가 더 상세하면 좋겠습니다. 지금은 `~~가 거부되었습니다.`와 같은 에러 메시지로 작성이 되어 있는데요, `~~하기 때문에 ~~가 거부되었습니다.`의 형태가 된다면 디버깅하기도 좀 더 편할 거에요.

Flying-Fisherman replied:
에러메세지도 상세하게 적도록 노력하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/security/UserAuthentication.java:
```diff
@@ -0,0 +1,55 @@
+
+        authorities.add(new SimpleGrantedAuthority("USER"));
+
+        if(userId == 999L) {
+            authorities.add(new SimpleGrantedAuthority("ADMIN"));
+        }
```
`999L`은 관리자인가요?

Flying-Fisherman replied:
네 999L로 아무 숫자나 넣어서 관리자로 하드코딩 했습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -25,6 +26,15 @@
 class UserControllerTest {
+    private static final String VALID_TOKEN = "eyJhbGciOiJIUzI1NiJ9." +
+            "eyJ1c2VySWQiOjF9.ZZ3CUl0jxeLGvQ1Js5nG2Ty5qGTlqai5ubDMXZOdaDk";
+    private static final String INVALID_TOKEN = VALID_TOKEN.replace("3", "4");
```
이 방식은 조금 조악한 것 같습니다.

johngrib reviewed (commented):
한 주간 고생 많으셨습니다. 주간 회고를 작성해 공유해 주세요.

Flying-Fisherman committed:
권한 부여 추가 및 관련 메서드 수정

`AuthenticationService`에서의 권한 부여 부분을 추가하였습니다.

`Role`, `JpaRoleRepository`, `RoleRepository`를 추가하였습니다.

`JwtAuthenticationFilter`에서의 권한 부여 부분을 추가하였습니다.

`ProductController`, `UserController`에서의 권한 확인부분을 추가하였습니다.

`UserAuthentication`에서의 권한 부여 부분을 추가하였습니다.

`UserService`에서의 수정하려는 계정과 수정하는 계정의 ID 확인부분의 오류를 수정하였습니다.

이 외 각 테스트에 대해 수정하였습니다.
